woocommerce/woocommerce-gutenberg-products-block
# REST API - Cart Order API

Implements REST endpoints to create orders from carts. The response heavily mimics the cart endpoint meaning the client can show details from the cart or the order in the same manner. 

Closes #1322 

## Cart Order API

Create a new order from the items in the cart.

```http
POST /cart/order/
```

| Attribute | Type   | Required | Description                                    |
| :-------- | :----- | :------: | :--------------------------------------------- |
| `billing_address`    | array |   No    | Billing address data to store to the new order. |
| `shipping_address`    | array |   No    | Shipping address data to store to the new order. |
| `customer_note`    | string |   No    | Customer note to store to the new order. |
| `shipping_rates`   | array  |    No    | Array of objects containing `rate_id` of selected shipping methods to add to the order. |

```http
curl --request POST https://example-store.com/wp-json/wc/store/cart/order
```

Example response:

<details>
<pre>
{
  "id": 149,
  "number": "149",
  "status": "draft",
  "order_key": "wc_order_9falc306dOkWb",
  "created_via": "store-api",
  "prices_include_tax": true,
  "events": {
    "date_created": "2020-01-07T12:33:23",
    "date_created_gmt": "2020-01-07T12:33:23",
    "date_modified": "2020-01-07T12:33:23",
    "date_modified_gmt": "2020-01-07T12:33:23",
    "date_completed": null,
    "date_completed_gmt": null,
    "date_paid": null,
    "date_paid_gmt": null
  },
  "customer": {
    "customer_id": 1,
    "customer_ip_address": "192.168.50.1",
    "customer_user_agent": "insomnia\/7.0.5"
  },
  "customer_note": "This is a customer note.",
  "billing_address": {
    "first_name": "Margaret",
    "last_name": "Thatchcroft",
    "company": "",
    "address_1": "123 South Street",
    "address_2": "Apt 1",
    "city": "Philadelphia",
    "state": "PA",
    "postcode": "19123",
    "country": "US",
    "email": "test@test.com",
    "phone": ""
  },
  "shipping_address": {
    "first_name": "Margaret",
    "last_name": "Thatchcroft",
    "company": "",
    "address_1": "123 South Street",
    "address_2": "Apt 1",
    "city": "Philadelphia",
    "state": "PA",
    "postcode": "19123",
    "country": "US"
  },
  "items": [
    {
      "id": 12,
      "quantity": 1,
      "name": "Belt",
      "sku": "woo-belt",
      "permalink": "http:\/\/local.wordpress.test\/product\/belt\/",
      "images": [
        {
          "id": "41",
          "src": "http:\/\/local.wordpress.test\/wp-content\/uploads\/2019\/12\/belt-2.jpg",
          "thumbnail": "http:\/\/local.wordpress.test\/wp-content\/uploads\/2019\/12\/belt-2-300x300.jpg",
          "srcset": "http:\/\/local.wordpress.test\/wp-content\/uploads\/2019\/12\/belt-2.jpg 801w, http:\/\/local.wordpress.test\/wp-content\/uploads\/2019\/12\/belt-2-300x300.jpg 300w, http:\/\/local.wordpress.test\/wp-content\/uploads\/2019\/12\/belt-2-100x100.jpg 100w, http:\/\/local.wordpress.test\/wp-content\/uploads\/2019\/12\/belt-2-450x450.jpg 450w, http:\/\/local.wordpress.test\/wp-content\/uploads\/2019\/12\/belt-2-150x150.jpg 150w, http:\/\/local.wordpress.test\/wp-content\/uploads\/2019\/12\/belt-2-768x768.jpg 768w",
          "sizes": "(max-width: 801px) 100vw, 801px",
          "name": "belt-2.jpg",
          "alt": ""
        }
      ],
      "variation": [],
      "totals": {
        "currency_code": "GBP",
        "currency_symbol": "£",
        "currency_minor_unit": 2,
        "currency_decimal_separator": ".",
        "currency_thousand_separator": ",",
        "currency_prefix": "£",
        "currency_suffix": "",
        "line_subtotal": "4583",
        "line_subtotal_tax": "917",
        "line_total": "4583",
        "line_total_tax": "917"
      }
    }
  ],
  "totals": {
    "currency_code": "GBP",
    "currency_symbol": "£",
    "currency_minor_unit": 2,
    "currency_decimal_separator": ".",
    "currency_thousand_separator": ",",
    "currency_prefix": "£",
    "currency_suffix": "",
    "total_items": "4583",
    "total_items_tax": "917",
    "total_fees": "0",
    "total_fees_tax": "0",
    "total_discount": "0",
    "total_discount_tax": "0",
    "total_shipping": "499",
    "total_shipping_tax": "100",
    "total_price": "6099",
    "total_tax": "1017",
    "tax_lines": [
      {
        "name": "Tax",
        "price": "1017"
      }
    ]
  }
}
</pre>
</details>

## Stock handling

Since we're moving to a system where orders are created upon checkout entry, we need to ensure stock is reserved during the checkout process. For this I've included a new table named `wc_reserved_stock` which stores order id, product id, and stock quantity. This table is then referenced during order creation to ensure there is enough stock to fulfil an order.

Stock reservations older than 10 mins or for non-draft orders are ignored. If this endpoint is called multiple times, stock reservations are renewed.

If there is insufficient stock available, the endpoint returns an error message and code `403`. We can use this to show a message and/or redirect back to the cart.

We could look into having a cleanup method for draft orders but that would likely fit into a followup. Right now drafts are harmless, but we'll need to chat about what we want doing with them if abandoned.

## To test

- Add some items to your cart whilst logged in
- Basic auth to API for the same user. 
- Do a post to `/store/cart/order`. 
- Check response.

---

There are some **follow-up items** from this that likely need looking at during the build of checkout:

- [x] **Chosen shipping options.** There is an inline todo. We need some way of recording (on the server side) which shipping option was chosen, and ensuring it matches up to presented options on checkout. I mention server side because we need to prevent tampering of costs making this unsuitable to pass via API calls.
- [ ] **Chosen payment method.** I think this needs to be set by the payment method when handling payment, but will need investigation.
- [ ] **Draft order timeouts.** Checkout should create an order on entry so stock is reserved. There needs to be something in place to ensure the draft order is kept-alive during the session. If inactive, the draft order may (and should be) deleted to free up stock for other customers.
